src/draive/scope/metrics.py

```diff
@@ -340,19 +339,51 @@ def log_debug(
 
     # - INTERNAL -
 
-    def _enter_task(self) -> None:
-        if self.is_finished:
-            raise ValueError("Attempting to use already finished metrics")
+    def enter_task(self) -> None:
+        assert not self.is_finished, "Attempting to use already finished metrics"  # nosec: B101
         self._pending_tasks += 1
 
-    def _exit_task(self) -> None:
+    def exit_task(  # noqa: PLR0912, C901
+        self,
+        exception: BaseException | None,
+    ) -> None:
+        assert not self.is_finished, "Attempting to use already finished metrics"  # nosec: B101
         assert self._pending_tasks > 0, "Unbalanced metrics task exit"  # nosec: B101
         self._pending_tasks -= 1
+        if self._exception is None:
+            self._exception = exception
+        elif isinstance(exception, Exception):
```

Comment:
It looks messy. Isn't `BaseException` parent to `Exception`? Same as `BaseExceptionGroup` is to `ExceptionGroup`?

Reply:
it is super messy but unfortunately may be required to keep the proper semantics, currently we are not using this exception besides logging so I might simplyfy it a bit